## Install dependencies as needed

In [3]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

!pip install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

# import json # library to handle JSON files
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
!pip install folium
import folium # map rendering library

print('Libraries imported.')


Usage:   
  pip install [options] <requirement specifier> [package-index-options] ...
  pip install [options] -r <requirements file> [package-index-options] ...
  pip install [options] [-e] <vcs project url> ...
  pip install [options] [-e] <local project path> ...
  pip install [options] <archive url/path> ...

no such option: --yes
     |████████████████████████████████| 102kB 7.7MB/s ta 0:00:011
Libraries imported.


## Build borough dataframe using process from prior parts of the activity

In [4]:
# Build borough dataframe using cells from first part of activity

import pandas as pd

url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df = pd.read_html(url)
# df contains a list of DataFrame objects, the first table on the web page is our Postal Code data, save that to a separate object
rawCodes = df[0]

assignedBoroughs = rawCodes
# remove the "Not assigned" entries in the Borough column
assignedBoroughs. drop(assignedBoroughs[assignedBoroughs["Borough"] == 'Not assigned'].index, axis=0, inplace=True)

# reset index
assignedBoroughs.reset_index(inplace=True, drop=True)

In [5]:
# read in lat/long data, merge with borough data to create input for our analysis
url = 'https://cocl.us/Geospatial_data'
coord_df = pd.read_csv(url)
Boroughs_coords = assignedBoroughs.merge(coord_df, how='inner', on='Postal Code', copy=False, validate='1:1')
Boroughs_coords.head(10)

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6         M1B       Scarborough                               Malvern, Rouge   
7         M3B        North York                                    Don Mills   
8         M4B         East York              Parkview Hill, Woodbine Gardens   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

## Connect to Foursquare API

#### Build connection string parameters
Note:  I will include my CLIENT ID and CLIENT SECRET in the next cell, and attempt to hide it to see if that works for Github publishing

In [27]:
# The code was removed by Watson Studio for sharing.

#### Identify boroughs containing the word "Toronto"

In [24]:
Toronto_df = Boroughs_coords[Boroughs_coords['Borough'].str.contains('Toronto')].reset_index(drop=True)
Toronto_df.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [25]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(Toronto_df['Borough'].unique()),
        Toronto_df.shape[0]
    )
)

The dataframe has 4 boroughs and 39 neighborhoods.


### Create a map of Toronto with neighborhoods superimposed on top.

#### Create an instance of a geopy user agent and get lat/long values for Toronto

In [17]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### build map of Toronto using latitude and longitude values

#### Note:  zoom_start = 12 seems to zoom closely enough and still contain all of the borough markers

In [30]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(Toronto_df['Latitude'], Toronto_df['Longitude'], Toronto_df['Borough'], Toronto_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Walk through process with first borough in the dataframe

### Get first borough's details

In [43]:
neighborhood_lat = Toronto_df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_long = Toronto_df.loc[0, 'Longitude'] # neighborhood longitude value
neighborhood_name = Toronto_df.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of the "{}" neighborhood are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_lat, 
                                                               neighborhood_long))

Latitude and longitude values of the "Regent Park, Harbourfront" neighborhood are 43.6542599, -79.3606359.


### Use Foursquare's API to get top 100 venues in this borough

#### Build request URL - connection parms are stored in a hidden cell

In [33]:
# Build URL
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, borough_lat, borough_long, VERSION, radius, LIMIT)

#### Issue request and examine results

In [34]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ef4caa095d86421fc12240f'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 44,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

In [35]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [38]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0          Roselle Desserts               Bakery  43.653447 -79.362017
1             Tandem Coffee          Coffee Shop  43.653559 -79.361809
2    Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3       Body Blitz Spa East                  Spa  43.654735 -79.359874
4  Dominion Pub and Kitchen                  Pub  43.656919 -79.358967

## Extend this process to find venues in all neighborhoods in our dataframe

### 1)  Build function to loop through all neighborhoods

In [44]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### 2)  Run the above function on each neighborhood and create a new dataframe, storead as *toronto_venues*.

In [45]:
toronto_venues = getNearbyVenues(names=Toronto_df['Neighborhood'],
                                   latitudes=Toronto_df['Latitude'],
                                   longitudes=Toronto_df['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

#### Check size and shape of toronto_venues

In [46]:
print(toronto_venues.shape)
toronto_venues.head()

(1620, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                      Venue  Venue Latitude  Venue Longitude  \
0          Roselle Desserts       43.653447       -79.362017   
1             Tandem Coffee       43.653559       -79.361809   
2    Cooper Koo Family YMCA       43.653249       -79.358008   
3       Body Blitz Spa East       43.654735       -79.359874   
4  Dominion Pub and Kitchen       43.656919       -79.358967   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4                  Pub

see how many venues were returned for each neighborhood

In [47]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
Brockton, Parkdale Village, Exhibition Place                           24   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     64   
Christie                                                               17   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           16   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         6   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         24   
Kensington Market, Chinatown, Grange Park                              62   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               43   
Moore Park, Summerhill East                                             3   
North Toronto West, Lawrence Park                                      20   
Parkdale, Roncesvalles                                                 15   
Queen's Park, Ontario Provincial Government                            34   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                               92   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     38   
St. James Town                                                         78   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         96   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             5   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
Brockton, Parkdale Village, Exhibition Place                            24   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      64   
Christie                                                                17   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

Note:  I noticed a couple of neighborhoods and a small number of venues listed, I will examine this outside of this exercise

#### How many unique categories came back from Foursquare?

In [48]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 unique categories.


### 3)  Analyze each neighborhood

#### Perform one-hot encoding on the categories

In [64]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood_Name'] = toronto_venues['Neighborhood'] 

# move neighborhood_name column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])

toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood_Name  Afghan Restaurant  Airport  Airport Food Court  \
0  Regent Park, Harbourfront                  0        0                   0   
1  Regent Park, Harbourfront                  0        0                   0   
2  Regent Park, Harbourfront                  0        0                   0   
3  Regent Park, Harbourfront                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       1     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Rental / Bike Share  Bistro  Board Shop  \
0                   0                         0       0           0   
1                   0                         0       0           0   
2                   0                         0       0           0   
3                   0                         0       0           0   
4                   0                         0       0           0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0     

In [65]:
toronto_onehot.shape

(1620, 238)

### Group rows by neighborhood and take the mean of the frequency of each category (for ranking purposes later)

In [67]:
toronto_grouped = toronto_onehot.groupby('Neighborhood_Name').mean().reset_index()
toronto_grouped

Neighborhood_Name  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.013333   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                  North Toronto West, Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0000              0.0000        0.0000           0.000   
3    0.0625              0.0625        0.0625           0.125   
4    0.0000              0.0000        0.0000           0.000   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000         

#### and confirm the new size - still should have 39 neighborhoods and 238 colums (neighborhood_name and 237 categories)

In [68]:
toronto_grouped.shape

(39, 238)

#### print each neighborhood with its top 5 categories

In [74]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood_Name']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood_Name'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2                Café  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.12
1     Coffee Shop  0.08
2  Breakfast Spot  0.08
3          Bakery  0.08
4     Yoga Studio  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
           venue  freq
0    Pizza Place  0.06
1  Auto Workshop  0.06
2     Restaurant  0.06
3        Butcher  0.06
4  Burrito Place  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.19
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3   Harbor / Marina  0.06
4           Airport  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.17
1       Sandwich Pla

### 4)  Load this data into a pandas dataframe for examination, mapping purposes

#### Build function to sort values in descending order

In [75]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Create dataframe with top-10 venues in each neighborhood

In [76]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood_Name']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...                Garden   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar              Beer Bar                Bakery   
1                Bakery           Coffee Shop        Breakfast Spot   
2               Brewery         Burrito Place                   Spa   
3        Airport Lounge      Airport Terminal      Sculpture Garden   
4        Sandwich Place    Italian Restaurant                  Café   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0           Cheese Shop    Seafood Restaurant                   Café   
1           Yoga Studio     Convenience Store  Performing Arts Venue   
2    Light Rail Station           Pizza Place                Butcher   
3               Airport    Airport Food Court           Airport Gate   
4   Japanese Restaurant           Salad Place        Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant      Greek Restaurant              Irish Pub  
1             Pet Store          Climbing Gym             Restaurant  
2  Fast Food Restaurant         Auto Workshop         Farmers Market  
3              Boutique         Boat or Ferry    Rental Car Location  
4          Burger Joint      Department Store        Thai Restaurant

### 5)  Cluster Neighborhoods

#### set up k-means clustering 

Note:   I started with five clusters and saw most neighborhoods lumped into the same cluster, 
        so I extended it to 10 to attempt to get more granularity in the results

        It didn't seem to help much

In [99]:
# set number of clusters
kclusters = 10

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood_Name', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:40] 

array([9, 9, 1, 4, 9, 1, 9, 9, 1, 7, 1, 9, 3, 9, 9, 1, 1, 1, 6, 9, 5, 9,
       9, 9, 9, 9, 2, 0, 1, 9, 1, 9, 1, 1, 1, 8, 1, 9, 1], dtype=int32)

Original attempt with 5 clusters - I see the clustering algorithm assigned most of the neighborhoods to cluster 3, hmmmm

I wonder if "Coffee Shop" and "Cafe" are driving the assignments

#### create a dataframe for the cluster and top-10 venues for each neighborhood

In [118]:
# add clustering labels
neighborhoods_venues_sorted10 = neighborhoods_venues_sorted  # create a copy for the 10-cluster list
neighborhoods_venues_sorted10.drop('Cluster Labels', axis=1, inplace=True)
neighborhoods_venues_sorted10.insert(0, 'Cluster Labels', kmeans.labels_)   # re-do since kmeans.labels_ now has 10 values

toronto_merged10 = Toronto_df

# merge toronto_merged with Toronto_df 
toronto_merged10 = toronto_merged10.join(neighborhoods_venues_sorted10.set_index('Neighborhood'), on='Neighborhood')

toronto_merged10.head(40) # check the last columns!

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         The Beaches  43.676357 -79.293031   
5                                         Berczy Park  43.644771 -79.373306   
6                                  Central Bay Street  43.657952 -79.387383   
7                                            Christie  43.669542 -79.422564   
8                            Richmond, Adelaide, King  43.650571 -79.384568   
9                        Dufferin, Dovercourt Village  43.669005 -79.442259   
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
11                           Little Portugal, Trinity  43.647927 -79.419750   
12                       The Danforth West, Riverdale  43.679557 -79.352188   
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
15                     India Bazaar, The Beaches West  43.668999 -79.315572   
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
17                                    Studio District  43.659526 -79.340923   
18                                      Lawrence Park  43.728020 -79.388790   
19                                           Roselawn  43.711695 -79.416936   
20                                   Davisville North  43.712751 -79.390197   
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
22                      High Park, The Junction South  43.661608 -79.464763   
23                  North Toronto West, Lawrence Park  43.715383 -79.405678   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
25                             Parkdale, Roncesvalles  43.648960 -79.456325   
26                                         Davisville  43.704324 -79.388790   
27                     University of Toronto, Harbord  43.662696 -79.400049   
28                                 Runnymede, Swansea  43.651571 -79.484450   
29                        Moore Park, Summerhill East  43.689574 -79.383160   
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
31  Summerhill West, Rathnelly, So

#### Let's visualize the clusters and popular venues

In [120]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged10['Latitude'], toronto_merged10['Longitude'], toronto_merged10['Neighborhood'], toronto_merged10['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### 6)  Let's look at each cluster to determine what could be driving the assignments

#### Cluster 1

In [123]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 0, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
19     Roselawn                Garden           Yoga Studio   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
19          Dance Studio  Eastern European Restaurant   Dumpling Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19            Donut Shop      Doner Restaurant               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
19   Distribution Center         Discount Store

#### Cluster 2

In [124]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 1, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue  \
7                                            Christie         Grocery Store   
9                        Dufferin, Dovercourt Village                Bakery   
12                       The Danforth West, Riverdale      Greek Restaurant   
15                     India Bazaar, The Beaches West                  Park   
17                                    Studio District                  Café   
22                      High Park, The Junction South                  Café   
24                The Annex, North Midtown, Yorkville                  Café   
26                                         Davisville           Pizza Place   
27                     University of Toronto, Harbord                  Café   
28                                 Runnymede, Swansea           Coffee Shop   
30          Kensington Market, Chinatown, Grange Park                  Café   
31  Summerhill West, Rathnelly, South Hill, Forest...                   Pub   
35                        St. James Town, Cabbagetown           Coffee Shop   
38  Business reply mail Processing Centre, South C...                Garden   

            2nd Most Common Venue   3rd Most Common Venue  \
7                            Café                    Park   
9                        Pharmacy               Pet Store   
12                    Coffee Shop      Italian Restaurant   
15           Fast Food Restaurant             Pizza Place   
17                    Coffee Shop               Gastropub   
22             Mexican Restaurant  Furniture / Home Store   
24                 Sandwich Place             Coffee Shop   
26                   Dessert Shop          Sandwich Place   
27            Japanese Restaurant                     Bar   
28                           Café             Pizza Place   
30  Vegetarian / Vegan Restaurant                  Bakery   
31                    Coffee Shop             Supermarket   
35                           Café                  Bakery   
38                        Brewery           Burrito Place   

    4th Most Common Venue   5th Most Common Venue      6th Most Common Venue  \
7                   Diner             Candy Store                 Baby Store   
9             Supermarket                    Café                        Bar   
12         Ice Cream Shop  Furniture / Home Store                 Restaurant   
15             Board Shop       Fish & Chips Shop                    Brewery   
17                 Bakery                 Brewery        American Restaurant   
22        Thai Restaurant              Restaurant                  Speakeasy   
24            Pizza Place                     Pub  Middle Eastern Restaurant   
26            Coffee Shop      Italian Restaurant                       Café   
27     Italian Restaurant               Bookstore                 Restaurant   
28       Sushi Restaurant                   Diner         Italian Restaurant   
30            Coffee Shop   Vietnamese Restaurant         Mexican Restaurant   
31  Vietnamese Restaurant        Sushi Restaurant         Light Rail Station   
35             Restaurant                    Park         Italian Restaurant   
38                    Spa      Light Rail Station                Pizza Place   

   7th Most Common Venue 8th Most Common Venue      9th Most Common Venue  \
7             Restaurant    Athletics & Sports         Italian Restaurant   
9       Recording Studio                  Bank  Middle Eastern Restaurant   
12             Bookstore       Bubble Tea Shop          Indian Restaurant   
15          Liquor Store            Restaurant              Burrito Place   
17          Neighborhood        Sandwich Place                Cheese Shop   
22           Flea Market                   Bar                     Bakery   
24             BBQ Joint     Indian Restaurant             History Museum   
26      Sushi Restaurant                   Gym                      Diner   
27                Bakery    Chinese Restauran

#### Cluster 3

In [125]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 2, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
33     Rosedale                  Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
33                 Trail          Cupcake Shop   Dumpling Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
33            Donut Shop      Doner Restaurant               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
33   Distribution Center         Discount Store

#### Cluster 4

In [126]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 3, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue  \
21  Forest Hill North & West, Forest Hill Road Park                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21                  Park              Bus Line         Jewelry Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21      Sushi Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21      Doner Restaurant               Dog Run    Distribution Center

#### Cluster 5

In [127]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 4, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue  \
32  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32        Airport Lounge      Airport Terminal      Sculpture Garden   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32               Airport    Airport Food Court          Airport Gate   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32              Boutique         Boat or Ferry    Rental Car Location

#### Cluster 6

In [128]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 5, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
29  Moore Park, Summerhill East                  Park            Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
29            Playground            Donut Shop      Doner Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
29               Dog Run   Distribution Center        Discount Store   

   9th Most Common Venue 10th Most Common Venue  
29                 Diner     Dim Sum Restaurant

#### Cluster 7

In [129]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 6, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
18  Lawrence Park                  Park              Bus Line   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18           Swim School                Lawyer  Colombian Restaurant   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
18           College Gym  Eastern European Restaurant   Dumpling Restaurant   

   9th Most Common Venue 10th Most Common Venue  
18            Donut Shop       Doner Restaurant

#### Cluster 8

In [130]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 7, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
20  Davisville North                  Park           Pizza Place   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
20        Breakfast Spot        Sandwich Place      Department Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
20     Food & Drink Shop     Convenience Store                 Hotel   

   9th Most Common Venue 10th Most Common Venue  
20                   Gym           Concert Hall

#### Cluster 9

In [131]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 8, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
4  The Beaches     Health Food Store          Neighborhood   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4      Asian Restaurant                   Pub                 Trail   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4      Department Store          Dessert Shop    Dim Sum Restaurant   

  9th Most Common Venue 10th Most Common Venue  
4                 Diner            Yoga Studio

#### Cluster 10

In [132]:
toronto_merged10.loc[toronto_merged10['Cluster Labels'] == 9, toronto_merged10.columns[[2] + list(range(6, toronto_merged10.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                           Regent Park, Harbourfront           Coffee Shop   
1         Queen's Park, Ontario Provincial Government           Coffee Shop   
2                            Garden District, Ryerson        Clothing Store   
3                                      St. James Town                  Café   
5                                         Berczy Park           Coffee Shop   
6                                  Central Bay Street           Coffee Shop   
8                            Richmond, Adelaide, King           Coffee Shop   
10  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
11                           Little Portugal, Trinity                   Bar   
13           Toronto Dominion Centre, Design Exchange           Coffee Shop   
14       Brockton, Parkdale Village, Exhibition Place                  Café   
16                     Commerce Court, Victoria Hotel           Coffee Shop   
23                  North Toronto West, Lawrence Park        Clothing Store   
25                             Parkdale, Roncesvalles             Gift Shop   
34                                     Stn A PO Boxes           Coffee Shop   
36             First Canadian Place, Underground city           Coffee Shop   
37                               Church and Wellesley           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0                   Park                   Pub                       Bakery   
1       Sushi Restaurant                 Diner                  Yoga Studio   
2            Coffee Shop                  Café               Cosmetics Shop   
3            Coffee Shop             Gastropub                   Restaurant   
5           Cocktail Bar              Beer Bar                       Bakery   
6         Sandwich Place    Italian Restaurant                         Café   
8                   Café            Restaurant                          Gym   
10              Aquarium                  Café                        Hotel   
11                  Café           Coffee Shop             Asian Restaurant   
13                 Hotel                  Café                   Restaurant   
14                Bakery           Coffee Shop               Breakfast Spot   
16                  Café            Restaurant                        Hotel   
23           Coffee Shop           Yoga Studio         Fast Food Restaurant   
25        Breakfast Spot         Movie Theater  Eastern European Restaurant   
34                  Café    Italian Restaurant           Seafood Restaurant   
36                  Café            Restaurant                        Hotel   
37      Sushi Restaurant   Japanese Restaurant                   Restaurant   

        5th Most Common Venue          6th Most Common Venue  \
0                        Café                        Theater   
1              Discount Store                            Bar   
2   Middle Eastern Restaurant            Japanese Restaurant   
3         American Restaurant                   Cocktail Bar   
5                 Cheese Shop             Seafood Restaurant   
6         Japanese Restaurant                    Salad Place   
8             Thai Restaurant                          Hotel   
10             Scenic Lookout                        Brewery   
11      Vietnamese Restaurant  Vegetarian / Vegan Restaurant   
13         Italian Restaurant             Seafood Restaurant   
14                Yoga Studio              Convenience Store   
16        American Restaurant                            Gym   
23        Rental Car Location             Mexican Restaurant   
25                    Dog Run             Italian Restaurant   
34                   Beer Bar                   Cocktail Bar   
36                        Gym                    Salad Place   
37                    Gay Bar                    Yoga Studio   

    7th Most Common Venue 8th Most Common 